cmd/common/st/transfer.go

```diff
@@ -0,0 +1,307 @@
+		if err != nil {
+			tracelog.ErrorLogger.FatalError(fmt.Errorf("can't configure target storage folder: %w", err))
+		}
+
+		parentDirPath := args[0]
+		err = transferFiles(sourceFolder, targetFolder, parentDirPath)
+		if err != nil {
+			tracelog.ErrorLogger.FatalError(err)
+		}
+	},
+}
+
+var (
+	transferSourceStorage            string
+	transferOverwrite                bool
+	transferConcurrency              int
+	transferMax                      int
+	transferAppearanceChecks         uint
+	transferAppearanceChecksInterval time.Duration
+)
+
+func init() {
+	transferCmd.Flags().StringVarP(&transferSourceStorage, "source", "s", "",
+		"storage name to move files from. Use 'default' to select the primary storage")
+	transferCmd.Flags().BoolVarP(&transferOverwrite, "overwrite", "o", false,
+		"whether to overwrite already existing files in the target storage and remove them from the source one")
+	transferCmd.Flags().IntVarP(&transferConcurrency, "concurrency", "c", 10,
+		"number of concurrent workers to move files. Value 1 turns concurrency off")
+	transferCmd.Flags().IntVarP(&transferMax, "max", "m", -1,
+		"max number of files to move in this run. Negative numbers turn the limit off")
+	transferCmd.Flags().UintVar(&transferAppearanceChecks, "appearance-checks", 3,
+		"number of times to check if a file is appeared for reading in the target storage after writing it. Value 0 turns checking off")
+	transferCmd.Flags().DurationVar(&transferAppearanceChecksInterval, "appearance-checks-interval", time.Second,
+		"minimum time interval between performing checks for files to appear in the target storage")
+
+	StorageToolsCmd.AddCommand(transferCmd)
+}
+
+func validateFlags() error {
+	if transferSourceStorage == "" {
+		return fmt.Errorf("source storage must be specified")
+	}
+	if transferSourceStorage == "all" {
+		return fmt.Errorf("an explicit source storage must be specified instead of 'all'")
+	}
+	if targetStorage == "all" {
+		return fmt.Errorf("an explicit target storage must be specified instead of 'all'")
+	}
+	if transferSourceStorage == targetStorage {
+		return fmt.Errorf("source and target storages must be different")
+	}
+	if transferConcurrency < 1 {
+		return fmt.Errorf("concurrency level must be >= 1 (which turns it off)")
+	}
+	return nil
+}
+
+func transferFiles(source, target storage.Folder, parentDirPath string) error {
+	files, err := listFilesToMove(source, target, parentDirPath)
+	if err != nil {
+		return err
+	}
+
+	fileStates := make(chan transferFileState, len(files))
+	for _, f := range files {
+		fileStates <- transferFileState{
+			path: f.GetName(),
+		}
+	}
+
+	workersNum := utility.Min(transferConcurrency, len(files))
+	errs := make(chan error, workersNum)
+
+	workersWG := new(sync.WaitGroup)
+	workersWG.Add(workersNum)
+	for i := 0; i < workersNum; i++ {
+		go moveFilesWorker(source, target, fileStates, errs, workersWG)
+	}
+
+	errsWG := new(sync.WaitGroup)
```

Comment:
use errorGroup

Reply:
It looks like `errroup.Wait` returns the first occurred error only. Is this an expected behaviour? In my initial implementation, we print all the errors, and exit with errorcode 1 only if there were any.

Reply:
> It looks like `errroup.Wait` returns the first occurred error only. Is this an expected behaviour? In my initial implementation, we print all the errors, and exit with errorcode 1 only if there were any.

Yes, if we encounter some critical one we will stop immediately and all other worker goroutines will get their context cancelled. We can handle non-critical errors inside the workers.